[GTK] Create WebKitWebExtension

ItsJamie9494 opened the pull request:
#### 785a4e3b033dd4e0c483fa961891d1e461123ff1
<pre>
[GTK] Create WebKitWebExtension
<a href="https://webkit.org/b/285379">https://webkit.org/b/285379</a>

Reviewed by Adrian Perez de Castro and Michael Catanzaro.

This creates a GLib API of WebExtension, along with some utility functions needed for Linux-specific pathways.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(WebKit::availableScreenScales):
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp: Added.
(webkit_web_extension_class_init):
(webkitWebExtensionCreate):
(webkitWebExtensionCreateWithBytes):
(webkit_web_extension_get_manifest_version):
(webkit_web_extension_supports_manifest_version):
(webkit_web_extension_get_default_locale):
(webkit_web_extension_get_display_name):
(webkit_web_extension_get_display_short_name):
(webkit_web_extension_get_display_version):
(webkit_web_extension_get_display_description):
(webkit_web_extension_get_display_action_label):
(webkit_web_extension_get_icon):
(webkit_web_extension_get_action_icon):
(webkit_web_extension_get_version):
(webkit_web_extension_get_requested_permissions):
(webkit_web_extension_get_optional_permissions):
(webkit_web_extension_get_requested_permission_match_patterns):
(webkit_web_extension_get_optional_permission_match_patterns):
(webkit_web_extension_get_all_requested_match_patterns):
(webkit_web_extension_get_has_background_content):
(webkit_web_extension_get_has_service_worker_background_content):
(webkit_web_extension_get_has_modular_background_content):
(webkit_web_extension_get_has_persistent_background_content):
(webkit_web_extension_get_has_injected_content):
(webkit_web_extension_get_has_options_page):
(webkit_web_extension_get_has_override_new_tab_page):
(webkit_web_extension_get_has_commands):
(webkit_web_extension_get_has_content_modification_rules):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionInternal.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::WebExtension):
* Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp: Added.
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::recordError):
(WebKit::WebExtension::iconForPath):
(WebKit::WebExtension::bestIcon):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtension.cpp: Added.
(testDisplayStringParsing):
(testDefaultLocaleParsing):
(testDisplayStringParsingWithLocalization):
(testActionParsing):
(testContentScriptsParsing):
(testPermissionsParsing):
(testBackgroundParsing):
(testBackgroundPreferredEnvironmentParsing):
(testOptionsPageParsing):
(testURLOverridesParsing):
(testContentSecurityPolicyParsing):
(testWebAccessibleResourcesV2):
(testWebAccessibleResourcesV3):
(testCommandsParsing):
(testDeclarativeNetRequestParsing):
(testExternallyConnectableParsing):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(TestWebKitAPI::Util::makePNGData):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.

Canonical link: <a href="https://commits.webkit.org/290957@main">https://commits.webkit.org/290957@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1aff1c770c2bf24059e98a187287f88e78d9f6f6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85361 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90489 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36403 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5185 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13073 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66361 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24174 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88407 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3956 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77531 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46643 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3838 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31792 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74554 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32630 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91965 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12708 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9296 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74932 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12937 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73368 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74051 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19429 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18438 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16866 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4729 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12651 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18119 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12481 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15974 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14232 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

GeorgesStavracas commented on Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp:
```diff
@@ -0,0 +1,60 @@
+    unsigned currentMonitor = 0;
+    while (auto item = g_list_model_get_item(monitors, currentMonitor++)) {
+        auto monitor = GDK_MONITOR(item);
+        if (monitor)
```
Is this check necessary? Gdk should guarantee that all items in the model are GdkMonitor objects, so probably better to `ASSERT(GDK_IS_MONITOR(item))` there, I guess.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -227,11 +258,14 @@ webkit_javascript_error_quark                  (void);
 
 WEBKIT_API GQuark
-webkit_user_content_filter_error_quark         (void);
+webkit_user_content_filter_error_quark          (void);
```
Looks like the whitespace here was nicer before.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.h.in:
```diff
@@ -0,0 +1,117 @@
+#ifndef WebKitWebExtension_h
+#define WebKitWebExtension_h
+
+#if ENABLE(WK_WEB_EXTENSIONS)
```
Remember you can't use conditionals in the public headers. These APIs should be available unconditionally.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.h.in:
```diff
@@ -0,0 +1,117 @@
+webkit_web_extension_get_has_commands                        (WebKitWebExtension *extension);
+
+WEBKIT_API gboolean
+webkit_web_extension_get_has_content_modification_rules       (WebKitWebExtension *extension);
```
```suggestion
webkit_web_extension_get_has_content_modification_rules      (WebKitWebExtension *extension);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionInternal.h:
```diff
@@ -0,0 +1,38 @@
+#include <wtf/HashMap.h>
+#include <wtf/text/WTFString.h>
+
+// Private API required by the Unit Tests
```
```suggestion
// Private API required by the unit tests
```

mcatanzaro commented on Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.cpp:
```diff
@@ -0,0 +1,61 @@
+    if (error)
+        return { };
+
+    return std::span { reinterpret_cast<const uint8_t*>(buffer.get()), bufferSize };
```
This will probably trip -Wunsafe-buffer-usage, but the solution is easy:

```suggestion
    return unsafeMakeSpan(reinterpret_cast<const uint8_t*>(buffer.get()), bufferSize);
```

mcatanzaro commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,187 @@
+        auto originalWidth = gdk_pixbuf_get_width(pixbuf.get());
+        auto originalHeight = gdk_pixbuf_get_height(pixbuf.get());
+        auto aspectWidth = sizeForResizing.width() / originalWidth;
+        auto aspectHeight = sizeForResizing.height() / originalHeight;
```
This is untrusted data, though. Width or height could be zero.

mcatanzaro commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,187 @@
+
+        gchar* buffer;
+        gsize* bufferSize;
+        gdk_pixbuf_save_to_buffer(pixbuf.get(), &buffer, bufferSize, "png"_s, NULL, NULL);
```
```suggestion
        gdk_pixbuf_save_to_buffer(pixbuf.get(), &buffer, bufferSize, "png"_s, nullptr, nullptr);
```

Are you sure it's OK to ignore errors here? I almost never ignore the error parameter since it makes debugging problems so much harder.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * 
+ * Since: 2.48
+ */
+gdouble webkit_web_extension_get_manifest_version(WebKitWebExtension *extension)
```
```suggestion
gdouble webkit_web_extension_get_manifest_version(WebKitWebExtension* extension)
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+        return nullptr;
+
+    priv->defaultLocale = defaultLocale.utf8();
+    return priv->defaultLocale.data();
```
OK, what's odd about all of the accessor functions in this file is you store the data in the priv struct so you have it handy in the future, but don't actually use it. You just reassign to the data each time, so there's no point in storing it in the first place.

The usual pattern is to check if it's already in the priv struct before reconstructing it:

```suggestion
    WebKitWebExtensionPrivate* priv = extension->priv;
    if (!priv->defaultLocale.isEmpty())
        return priv->defaultLocale.data();
    
    auto defaultLocale = priv->extension->defaultLocale();
    if (defaultLocale.isEmpty())
        return nullptr;

    priv->defaultLocale = defaultLocale.utf8();
    return priv->defaultLocale.data();
```

Same applies to all the other accessor functions.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * 
+ * Since: 2.48
+ */
+const gchar* webkit_web_extension_get_default_locale(WebKitWebExtension *extension)
```
```suggestion
const gchar* webkit_web_extension_get_default_locale(WebKitWebExtension* extension)
```

You've copy/pasted this style error throughout this file.

mcatanzaro reviewed (approved):
As before, this requires two GTK/WPE reviewers. But here's one!

carlosgcampos commented on Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp:
```diff
@@ -0,0 +1,60 @@
+#if USE(GTK4)
+    auto* monitors = gdk_display_get_monitors(display);
+    unsigned currentMonitor = 0;
+    while (auto item = g_list_model_get_item(monitors, currentMonitor++)) {
```
```suggestion
    while (auto* item = g_list_model_get_item(monitors, currentMonitor++)) {
```

carlosgcampos commented on Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp:
```diff
@@ -0,0 +1,60 @@
+    auto* monitors = gdk_display_get_monitors(display);
+    unsigned currentMonitor = 0;
+    while (auto item = g_list_model_get_item(monitors, currentMonitor++)) {
+        auto monitor = GDK_MONITOR(item);
```
```suggestion
        auto* monitor = GDK_MONITOR(item);
```

carlosgcampos commented on Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp:
```diff
@@ -0,0 +1,60 @@
+    }
+#endif
+
+    if (screenScales.size())
```
```suggestion
    if (!screenScales.isEmpty())
```

carlosgcampos commented on Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp:
```diff
@@ -0,0 +1,60 @@
+#include "config.h"
+#include "WebExtensionUtilities.h"
+
+#include <gdk/gdk.h>
```
This file is WebExtensionUtilitiesGlib (we normally prefer GLib), but it's using gdk unconditionally. If this is supposed to be gtk specific, use WebExtensionsUtilitiesGtk instead for the file name, otherwise use `#if PLATFORM(GTK)`

carlosgcampos commented on Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp:
```diff
@@ -0,0 +1,60 @@
+
+namespace WebKit {
+
+Vector<double> availableScreenScales()
```
Where is this used? where is the header of this file?

ItsJamie9494 replied:
`Shared/Extensions/WebExtensionUtilities.h` is the header. The function is used in `UIProcess/Extensions/WebExtension.cpp`

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+        auto string = resources.get(key);
+        auto data = API::Data::create(string.utf8().span());
+
+        newResources.set(key, data);
```
```suggestion
    for (const auto& it : resources)
        newResources.add(it.key, API::Data::create(it.value.utf8().span()));
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+    Ref extension = WebKit::WebExtension::create(WTFMove(newResources));
+
+    if (!extension->errors().isEmpty()) {
+        auto internalError = extension->errors().last().ptr();
```
```suggestion
        auto* internalError = extension->errors().last().ptr();
```

carlosgcampos replied:
Why do you need the ptr?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+WebKitWebExtension* webkitWebExtensionCreateWithBytes(const HashMap<String, std::span<const uint8_t>>& resources, GError** error)
+{
+    HashMap<String, Ref<API::Data>> newResources;
+    for (auto key : resources.keys()) {
```
Same here you can get the key value pair iterator.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * The returned image will be the best match for the specified size that is available in the extension's action icon set. If no matching icon is available,
+ * the method will fall back to the extension's icon.
+ * 
+ * Returns: the icon image, or %NULL if no icon could be loaded.
```
This is transfer none, I guess?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * 
+ * Since: 2.48
+ */
+GIcon* webkit_web_extension_get_action_icon(WebKitWebExtension *extension, gdouble width, gdouble height)
```
```suggestion
GIcon* webkit_web_extension_get_action_icon(WebKitWebExtension* extension, gdouble width, gdouble height)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * 
+ * Since: 2.48
+ */
+const gchar* webkit_web_extension_get_version(WebKitWebExtension *extension)
```
```suggestion
const gchar* webkit_web_extension_get_version(WebKitWebExtension* extension)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+
+    GList* returnValue = nullptr;
+    for (Ref pattern : matchPatternSet)
+        returnValue = g_list_prepend(returnValue, webkit_web_extension_match_pattern_ref(webkitWebExtensionMatchPatternCreate(pattern)));
```
Why do you need to ref a newly created pattern?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * 
+ * Since: 2.48
+ */
+GList* webkit_web_extension_get_optional_permission_match_patterns(WebKitWebExtension *extension)
```
```suggestion
GList* webkit_web_extension_get_optional_permission_match_patterns(WebKitWebExtension* extension)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * 
+ * Since: 2.48
+ */
+GList* webkit_web_extension_get_all_requested_match_patterns(WebKitWebExtension *extension)
```
```suggestion
GList* webkit_web_extension_get_all_requested_match_patterns(WebKitWebExtension* extension)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+ * 
+ * Since: 2.48
+ */
+gboolean webkit_web_extension_get_has_background_content(WebKitWebExtension *extension)
```
```suggestion
gboolean webkit_web_extension_get_has_background_content(WebKitWebExtension* extension)
```

carlosgcampos replied:
And everywhere else

carlosgcampos commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,187 @@
+    m_resources.set("manifest.json"_s, API::Data::create(manifestData.utf8().span()));
+}
+
+
```
```suggestion
```

carlosgcampos commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,187 @@
+
+    // Remove leading slash to normalize the path for lookup/storage in the cache dictionary.
+    if (path.startsWith('/'))
+        path = path.substring(1);
```
This has already been done

carlosgcampos commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,187 @@
+    if (imageData->span().empty())
+        return nullptr;
+
+    auto gimageBytes = adoptGRef(g_bytes_new(imageData->span().data(), imageData->size()));
```
```suggestion
    auto* gimageBytes = adoptGRef(g_bytes_new(imageData->span().data(), imageData->size()));
```

carlosgcampos commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,187 @@
+        gsize* bufferSize;
+        gdk_pixbuf_save_to_buffer(pixbuf.get(), &buffer, bufferSize, "png"_s, NULL, NULL);
+
+        gimageBytes = g_bytes_new(buffer, *bufferSize);
```
This is leaking the buffer, use g_bytes_new_take instead

carlosgcampos commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,187 @@
+
+    auto* image = g_bytes_icon_new(gimageBytes.get());
+
+    return WebCore::Icon::create(image);
```
WebCore::Icon::create() should receive a GRefPtr&& instead of pointer to make it clear it's taking the ownership

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1122 @@
+    g_assert_null(webkit_web_extension_get_version(extension));
+    g_assert_error(error.get(), WEBKIT_WEB_EXTENSION_ERROR, WEBKIT_WEB_EXTENSION_ERROR_INVALID_MANIFEST_ENTRY);
+
+    extension = parse("{ \"manifest_version\": 2, \"name\": \"Test\", \"version\": \"1.0\", \"description\": \"Test description\" }");
```
This is leaking the previous extension, you should use GRefPtr and adoptGRef

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1023 @@
+
+    WebKitWebExtension* object = WEBKIT_WEB_EXTENSION(g_object_new(WEBKIT_TYPE_WEB_EXTENSION, nullptr));
+    object->priv->extension = WTFMove(extension);
+    return object;
```
Do we really need two create functions? They are mostly the same thing. Maybe we could have a single one that receives GBytes instead of String/span and add an API::Data constructor taking a GRefPtr<GBytes> so that we also avoid copies.

aperezdc commented on Source/WebKit/Shared/API/APIData.h:
```diff
@@ -76,6 +81,17 @@ class Data : public ObjectImpl<API::Object::Type::Data> {
+        const void* rawData = g_bytes_get_data(bytes.get(), &size);
+
+        auto data = unsafeMakeSpan(static_cast<const uint8_t*>(rawData), size);
+        return createWithoutCopying(data, [bytes = WTFMove(bytes)] () { });
```
We have a `span(const GRefPtr<GBytes&>)` helper in `<wtf/glib/GSpanExtras.h>` which would be perfect to use here:

```suggestion
        return createWithoutCopying(span(bytes), [bytes = WTFMove(bytes)] () { });
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
```diff
@@ -155,6 +156,33 @@ unsigned toWebKitError(unsigned webCoreError)
+    case WebKit::WebExtension::APIError::Unknown:
+    default:
+        return WEBKIT_WEB_EXTENSION_ERROR_UNKNOWN;
+    }
```
I would remove the `default:` case and let the compiler check that every possible value of the `enum` is being handled:

```suggestion
        return WEBKIT_WEB_EXTENSION_ERROR_UNKNOWN;
    }
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1156 @@
+    N_PROPERTIES,
+};
+
+static GParamSpec* properties[N_PROPERTIES] = { nullptr, };
```
Using `std::array` here gives us bounds checking and better compiler diagnostics:

```suggestion
static std::array<GParamSpec*, N_PROPERTIES> properties;
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1156 @@
+            FALSE,
+            WEBKIT_PARAM_READABLE);
+
+    g_object_class_install_properties(objectClass, N_PROPERTIES, properties);
```
...and then here we use the `std::array` accessors to get the pointer and number of elements:
```suggestion
    g_object_class_install_properties(objectClass, properties.size(), properties.data());
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1156 @@
+ *
+ * Get the default locale for the extension.
+ * 
+ * Returns: the default locale, or %NULL if there was no default
```
This may return `NULL`:

```suggestion
 * Returns: (nullable): the default locale, or %NULL if there was no default
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
```diff
@@ -0,0 +1,1156 @@
+ *
+ * Get the localized name for the extension.
+ * 
+ * Returns: the localized name, or %NULL if there was no
```
Same here, this needs a `(nullable)` annotation. By default only `gpointer` and closure data parameters are considered nullable ([docs](https://gi.readthedocs.io/en/latest/annotations/giannotations.html#nullable-parameters)). Please do check the rest of functions and add the annotation where needed.

aperezdc commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -0,0 +1,194 @@
+
+    String path = originalPath;
+
+    // Remove leading slash to normalize the path for lookup/storage in the cache dictionary.
+    if (path.startsWith('/'))
+        path = path.substring(1);
```
Can we get a string here with many leading slashes? If so, it would be better to remove them all.

ItsJamie9494 replied:
I doubt it, and Cocoa's implementation doesn't seem to be concerned about it either as they only remove the one. The other similar function (resourceStringForPath in WebExtension.cpp) only handles the case of one leading slash.

aperezdc replied:
Let's then not worry about this for now. Thanks for commenting!

aperezdc reviewed (commented):
LGTM overall, I left a few small comments that you may want to address before landing.

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/IconGtk.cpp:
```diff
@@ -44,12 +44,12 @@ Icon::~Icon()
         return nullptr;
 
-    return adoptRef(new Icon(icon));
+    return adoptRef(new Icon(icon.get()));
```
You should change the constructor too, to receive a GRefPtr&& and here WTFMove(icon) instead of passing the raw pointer

carlosgcampos commented on Source/WebKit/Shared/API/APIData.h:
```diff
@@ -76,6 +82,14 @@ class Data : public ObjectImpl<API::Object::Type::Data> {
 #endif
 
+#if PLATFORM(GTK)
+    static Ref<Data> createWithoutCopying(GRefPtr<GBytes> bytes)
```
```suggestion
    static Ref<Data> createWithoutCopying(GRefPtr<GBytes>&& bytes)
```

ItsJamie9494 committed:
[GTK] Create WebKitWebExtension
https://webkit.org/b/285379

Reviewed by Adrian Perez de Castro and Michael Catanzaro.

This creates a GLib API of WebExtension, along with some utility functions needed for Linux-specific pathways.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Shared/Extensions/glib/WebExtensionUtilitiesGlib.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(WebKit::availableScreenScales):
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp: Added.
(webkit_web_extension_class_init):
(webkitWebExtensionCreate):
(webkitWebExtensionCreateWithBytes):
(webkit_web_extension_get_manifest_version):
(webkit_web_extension_supports_manifest_version):
(webkit_web_extension_get_default_locale):
(webkit_web_extension_get_display_name):
(webkit_web_extension_get_display_short_name):
(webkit_web_extension_get_display_version):
(webkit_web_extension_get_display_description):
(webkit_web_extension_get_display_action_label):
(webkit_web_extension_get_icon):
(webkit_web_extension_get_action_icon):
(webkit_web_extension_get_version):
(webkit_web_extension_get_requested_permissions):
(webkit_web_extension_get_optional_permissions):
(webkit_web_extension_get_requested_permission_match_patterns):
(webkit_web_extension_get_optional_permission_match_patterns):
(webkit_web_extension_get_all_requested_match_patterns):
(webkit_web_extension_get_has_background_content):
(webkit_web_extension_get_has_service_worker_background_content):
(webkit_web_extension_get_has_modular_background_content):
(webkit_web_extension_get_has_persistent_background_content):
(webkit_web_extension_get_has_injected_content):
(webkit_web_extension_get_has_options_page):
(webkit_web_extension_get_has_override_new_tab_page):
(webkit_web_extension_get_has_commands):
(webkit_web_extension_get_has_content_modification_rules):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionInternal.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::WebExtension):
* Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp: Added.
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::recordError):
(WebKit::WebExtension::iconForPath):
(WebKit::WebExtension::bestIcon):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtension.cpp: Added.
(testDisplayStringParsing):
(testDefaultLocaleParsing):
(testDisplayStringParsingWithLocalization):
(testActionParsing):
(testContentScriptsParsing):
(testPermissionsParsing):
(testBackgroundParsing):
(testBackgroundPreferredEnvironmentParsing):
(testOptionsPageParsing):
(testURLOverridesParsing):
(testContentSecurityPolicyParsing):
(testWebAccessibleResourcesV2):
(testWebAccessibleResourcesV3):
(testCommandsParsing):
(testDeclarativeNetRequestParsing):
(testExternallyConnectableParsing):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.cpp: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.
(TestWebKitAPI::Util::makePNGData):
* Tools/TestWebKitAPI/glib/WebKitGLib/WebExtensionUtilities.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitPrivate.h.

Canonical link: https://commits.webkit.org/290957@main
